MParT/MultivariateExpansionWorker.h

```diff
@@ -454,6 +416,136 @@ class MultivariateExpansionWorker
+
+        if constexpr (!std::is_same_v<Rectifier, Identity>) {
+            termVal = Rectifier::Evaluate(termVal);
+        }
+        termVal *= diagVal;
+        return termVal;
+    }
+
+    KOKKOS_FUNCTION double GetTermValInputDeriv(unsigned int termInd, int wrt, const double* polyCache) const {
+        double termVal = 1.0; // Compute the value of this term in the expansion
+        double wrtVal = 1.; // value of univariate basis fcn in dim wrt
+        double wrtDeriv = (wrt == -1); // derivative of univariate basis fcn in dim wrt
+        unsigned int end_idx = multiSet_.nzStarts(termInd+1)-1; // last index in loop
+        for(unsigned int i=multiSet_.nzStarts(termInd); i < end_idx; ++i){
+            if(int(multiSet_.nzDims(i))==wrt){
+                // wrt is the dimension we diff with respect to
+                wrtVal = polyCache[startPos_(wrt) + multiSet_.nzOrders(i)];
+                // We add dim_ to get the derivative index
+                wrtDeriv = polyCache[startPos_(dim_+wrt) + multiSet_.nzOrders(i)];
+            }else{
+                termVal *= polyCache[startPos_(multiSet_.nzDims(i)) + multiSet_.nzOrders(i)];
+            }
+        }
+        int d = multiSet_.nzDims(end_idx); // Index of last nonconstant dimension
+        double lastVal;
+        if(wrt == d) {
+            lastVal = 1;
+            wrtVal = polyCache[startPos_(d) + multiSet_.nzOrders(end_idx)];
+            wrtDeriv = polyCache[startPos_(dim_+d) + multiSet_.nzOrders(end_idx)];
+        } else {
+            lastVal = polyCache[startPos_(d) + multiSet_.nzOrders(end_idx)];
+        }
+        if constexpr (!std::is_same_v<Rectifier, Identity>) {
+            if(wrt == dim_ - 1) { // Diagonal deriv
+                termVal = Rectifier::Evaluate(termVal)*wrtDeriv; // if wrt != d, wrtDeriv = 0
+            } else if (wrt == -1) { // No deriv
+                termVal = (d == dim_ - 1) ? Rectifier::Evaluate(termVal)*lastVal : Rectifier::Evaluate(termVal*lastVal);
+            } else { // Offdiag deriv
+                if(d != dim_ - 1) wrtVal *= lastVal; // lastVal belongs on inside and outside
+                termVal = Rectifier::Derivative(termVal*wrtVal)*termVal*wrtDeriv*lastVal;
+            }
+        } else { // Reduce to loop body
+            termVal *= lastVal*wrtDeriv;
+        }
+        return termVal;
+    }
+
+    KOKKOS_FUNCTION double GetTermValMixedDeriv(unsigned int termInd, int wrt, const double* polyCache) const {
+        // Compute the value of this term in the expansion
+        double termVal = 1.0;
+        double wrtVal = 1.; // value of univariate basis fcn in dim wrt
+        double wrtDeriv = (wrt == int(dim_)-1 || wrt == -1); // derivative of univariate basis fcn in dim wrt
+        unsigned int end_idx = multiSet_.nzStarts(termInd+1)-1; // last index in loop
+        
+        if(multiSet_.nzDims(end_idx)!=dim_-1) return 0.; // Value is zero if constant in last dimension
```

Comment:
Same concern about order dependence.  Maybe we should just make sure in the compressed multiindex that the nonzero dimensions are sorted.